File: haskell/day19b.hs
import qualified Data.Either as E
import qualified Data.Map as Map
import qualified Data.Maybe as Maybe
import qualified Text.ParserCombinators.Parsec as PCPS

data RuleItem =
    SubRule Int |
    LiteralString String
    deriving (Show)

eol = PCPS.char '\n'

inputRuleItemSubRuleNum = do
    val <- PCPS.many1 PCPS.digit
    return $ SubRule (read val :: Int)

inputRuleItemString = do
    PCPS.char '"'
    s <- PCPS.many1 (PCPS.noneOf "\"")
    PCPS.char '"'
    return $ LiteralString s

inputRuleItem = do
    PCPS.char ' '
    item <- inputRuleItemSubRuleNum PCPS.<|> inputRuleItemString
    return item

inputRule = do
    ruleNum <- PCPS.many1 PCPS.digit
    PCPS.char ':'
    options <- PCPS.sepBy (PCPS.many1 $ PCPS.try inputRuleItem) (PCPS.string " |")
    eol
    return (read ruleNum :: Int, options)

inputMessage = do
    s <- PCPS.many (PCPS.noneOf "\n")
    eol
    return s

inputFile = do
    rules <- PCPS.many inputRule
    eol
    messages <- PCPS.many inputMessage
    PCPS.eof
    return (rules, messages)

matchRuleList :: Map.Map Int [[RuleItem]] -> String -> Int -> [RuleItem] -> [Int]
matchRuleList rules s i [] = [i]
matchRuleList rules s i ruleList =
    foldl (++) [] [matchRuleList rules s si (tail ruleList) | si <- matchRule rules s i (ruleNum $ head ruleList)]
    where
        ruleNum (SubRule n) = n

matchRule :: Map.Map Int [[RuleItem]] -> String -> Int -> Int -> [Int]
matchRule rules s i ruleNum =
    foldl (++) [] [subMatches ruleList | ruleList <- Maybe.fromJust $ Map.lookup ruleNum rules]
    where
        subMatches [(LiteralString s')] =
            if i < (length s) && s !! i == s' !! 0 then [i + 1] else []
        subMatches ruleList =
            matchRuleList rules s i ruleList

isMatch :: Map.Map Int [[RuleItem]] -> String -> Bool
isMatch rules s =
    (length s) `elem` (matchRule rules s 0 0)

answer :: ([(Int, [[RuleItem]])], [String]) -> Int
answer (rulesList, messages) = countMatch
    where
        rulesOrig = Map.fromList rulesList
        modifiedRules = Map.fromList [
            (8,  [[SubRule 42], [SubRule 42, SubRule 8]]),
            (11, [[SubRule 42, SubRule 31], [SubRule 42, SubRule 11, SubRule 31]])]
        rules = Map.union modifiedRules rulesOrig
        isMatches = map (isMatch rules) messages
        countMatch = length $ filter (== True) $ isMatches

main :: IO ()
main = do
    let fn = "../input/day19.txt"
    inputText <- readFile fn
    let input = E.fromRight ([], []) $ PCPS.parse inputFile fn inputText
    putStrLn $ show $ answer $ input

File: python/day19b.py
#!/usr/bin/env python3

rules = {}
messages = []
def parse_item(s):
    if s[0] == '"':
        return s[1]
    return int(s)
with open("../input/day19.txt") as f:
    # 117: "b"
    # 68: 33 99 | 117 35
    while True:
        l = f.readline().strip()
        if not l:
            break
        (s_rnum, s_options) = l.split(': ')
        ls_options = s_options.split(' | ')
        options = []
        for s_option in ls_options:
            options.append(list(map(parse_item, s_option.split())))
        rules[int(s_rnum)] = options
    # abbbaaaaaabbbaabababaaab
    while True:
        l = f.readline().strip()
        if not l:
            break
        messages.append(l)

rules[8] = [[42], [42, 8]]
rules[11] = [[42, 31], [42, 11, 31]]

def match_rule_list(s, i, rule_list, indent):
    #print(indent + f'match_rule_list(s={s}, i={i}, rule_list={rule_list}, sleft={s[i:]})')
    if not rule_list:
        yield i
        return
    for si in match_rule(s, i, rule_list[0], indent):
        yield from match_rule_list(s, si, rule_list[1:], indent)

def match_rule(s, i, ruleno, indent):
    #print(indent + f'match_rule(s={s}, i={i}, ruleno={ruleno}, sleft={s[i:]})')
    for rule_list in rules[ruleno]:
        #print(indent + f'  rule_list={rule_list}')
        if isinstance(rule_list[0], str):
            if i < len(s) and s[i] == rule_list[0]:
                yield i + 1
            return
        yield from match_rule_list(s, i, rule_list, indent + '  ')

def is_match(s):
    return len(s) in match_rule(s, 0, 0, '')

count = 0
for s in messages:
    m = is_match(s)
    #print(s, m)
    if m:
        count += 1
print(count)
